File: ProLog Play/8_MergeSort.pl
% 8 Questao
mergesort( [], [] ).
mergesort( [X], [X] ).
mergesort( U, S ) :- divide(U, L, R), mergesort(L, SL), mergesort(R, SR), merge(SL, SR, S).

divide( [], [], [] ).
divide( [X], [X], [] ).
divide( [L,R|T], [L|LT], [R|RT] ) :- divide( T, LT, RT ).

merge( [], RS, RS ).
merge( LS, [], LS ).
merge( [L|LS], [R|RS], [L|T] ) :- L =< R, merge(    LS, [R|RS], T).
merge( [L|LS], [R|RS], [R|T] ) :- L > R,  merge( [L|LS],   RS,  T).

File: Haskell Play/8_MergeSort.hs
-- Questao 8
primeira_metade :: [a] -> [a]
primeira_metade xs = take (length xs `div` 2) xs

segunda_metade :: [a]->[a]
segunda_metade xs = drop (length xs `div` 2) xs

merge :: Ord a => [a]->[a]->[a]
merge xs [] = xs
merge [] ys = ys
merge (x:xs) (y:ys)
	|(x <= y) = x:(merge xs (y:ys))
	|otherwise = y:(merge(x:xs)ys)

mergesort :: Ord a => [a] -> [a]
mergesort [] = []
mergesort [x] = [x]
mergesort xs = merge (mergesort (primeira_metade xs)) (mergesort (segunda_metade xs))
